gateway-js/src/datasources/types.ts

```diff
@@ -1,7 +1,33 @@
 import { GraphQLResponse, GraphQLRequestContext } from 'apollo-server-types';
 
-export interface GraphQLDataSource<TContext extends Record<string, any> = Record<string, any>> {
+export interface GraphQLDataSource<
+  TContext extends Record<string, any> = Record<string, any>,
+> {
   process(
-    request: Pick<GraphQLRequestContext<TContext>, 'request' | 'context'>,
+    options: GraphQLDataSourceProcessOptions<TContext>,
   ): Promise<GraphQLResponse>;
 }
+
+export interface GraphQLDataSourceProcessOptions<
+  TContext extends Record<string, any> = Record<string, any>,
+> {
+  /**
+   * The request to send to the subgraph.
+   */
+  request: GraphQLRequestContext<TContext>['request'];
+  /**
+   * The GraphQLRequestContext for the operation received by the gateway, or
+   * one of the strings if this operation is generated by the gateway without an
+   * incoming request.
+   */
+  topLevelRequestContext:
```

Comment:
@trevor-scheer Do you like this name?

Do you like this type or should we go with something more like `{kind: 'health check'}|{kind: 'loading schema'}|{kind: 'incoming operation', requestContext: GraphQLRequestContext}`

Reply:
OK, I think we definitely want the `kind` sort of thing in case we want more information associated with health checks etc, so I've rebased the PR so it does that. Still curious if you like the topLevelRequestContext name :) I think it's better than incomingRequestContext because not all are based on incoming requests. Of course the fact that it doesn't always have a requestContext on it is a little odd. 🤷